### DISCUSSION TITLE
Delete propagation

### DISCUSSION THREAD

**Discussion by DannyBen** [soveran/ohm#248] [Q&A]
I was under the assumptions that deleting a parent object, deletes any object that references it.
The README only mentions this propagation in the [`track` section](https://github.com/soveran/ohm#tracked-keys).

I was expecting the below test to work - am I doing something wrong, or do I need to delete the orphan objects manually?

```ruby
require 'ohm'

class Parent < Ohm::Model
  attribute :name
  collection :children, :Child
end

class Child < Ohm::Model
  attribute :name
  reference :parent, :Parent
end

system 'redis-cli flushall'  # WARNING
parent = Parent.create name: 'Robert'
child = Child.create parent: parent, name: 'Kim'

parent.delete

p Parent[1].name rescue 'Parent is gone (PASS)'
p Child[1].name rescue 'Child is gone (PASS)'
p Child[1].parent

p Parent.all.count
p Child.all.count

```

**Comment by scalp42** [comment]
@DannyBen any updates by any chance? Curious how you ended working around this? Thanks in advance for sharing.

**Comment by DannyBen** [comment]
I don;t recall. Looking at my code I see nothing that handles this other than a function that deletes orphan objects...

**Comment by soveran** [comment]
Maybe we can use [Ohm::Callbacks](https://github.com/cyx/ohm-contrib/blob/master/lib/ohm/callbacks.rb).

By defining a `before_delete` callback, we can remove all the children:

```ruby
class Parent < Ohm::Model
  include Ohm::Callbacks

  attribute :name
  collection :children, :Child

  def before_delete
    children.each do |child|
      child.delete
    end
  end
end
```

**Reply by scalp42** [reply]
Ah very neat, merci @soveran!
